File: app/src/main/java/com/studenthub/app/model/ChatMessage.java
package com.studenthub.app.model;

import com.google.firebase.firestore.Exclude;
import com.google.firebase.firestore.PropertyName;
import com.google.firebase.firestore.ServerTimestamp;

import java.util.Date;

/**
 * Maps directly to a document under Chats/{roomId}/Messages in Firestore —
 * the SAME collection path and field names the Web App (app.js) uses. That
 * shared schema is what makes chat cross-platform: a message written here
 * shows up on the website instantly, and vice-versa.
 */
public class ChatMessage {

    private String senderId;
    private String senderName;
    private String senderPhoto;
    private String text;

    @ServerTimestamp
    private Date createdAt;

    private boolean edited;

    @Exclude
    private String id; // Firestore document id, filled in after reading

    public ChatMessage() {
        // Required empty constructor for Firestore deserialization
    }

    public ChatMessage(String senderId, String senderName, String senderPhoto, String text) {
        this.senderId = senderId;
        this.senderName = senderName;
        this.senderPhoto = senderPhoto;
        this.text = text;
    }

    public String getSenderId() { return senderId; }
    public void setSenderId(String senderId) { this.senderId = senderId; }

    public String getSenderName() { return senderName; }
    public void setSenderName(String senderName) { this.senderName = senderName; }

    public String getSenderPhoto() { return senderPhoto; }
    public void setSenderPhoto(String senderPhoto) { this.senderPhoto = senderPhoto; }

    public String getText() { return text; }
    public void setText(String text) { this.text = text; }

    public Date getCreatedAt() { return createdAt; }
    public void setCreatedAt(Date createdAt) { this.createdAt = createdAt; }

    @PropertyName("edited")
    public boolean isEdited() { return edited; }
    @PropertyName("edited")
    public void setEdited(boolean edited) { this.edited = edited; }

    @Exclude
    public String getId() { return id; }
    @Exclude
    public void setId(String id) { this.id = id; }
}

File: android/app/src/main/java/com/studenthub/app/data/model/ChatMessage.kt
package com.studenthub.app.data.model

import com.google.firebase.Timestamp

data class ReplyTo(
    val senderName: String = "",
    val text: String = ""
)

data class ChatMessage(
    val id: String = "",
    val text: String = "",
    val photos: List<String> = emptyList(),
    val photoUrl: String = "",
    val senderId: String = "",
    val senderName: String = "",
    val senderPhoto: String = "",
    val createdAtMillis: Long = 0L,
    val replyTo: ReplyTo? = null
) {
    companion object {
        fun fromMap(id: String, map: Map<String, Any?>): ChatMessage {
            val replyMap = map["replyTo"] as? Map<*, *>
            val reply = replyMap?.let {
                ReplyTo(
                    senderName = it["senderName"] as? String ?: "",
                    text = it["text"] as? String ?: ""
                )
            }
            return ChatMessage(
                id = id,
                text = map["text"] as? String ?: "",
                photos = (map["photos"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
                photoUrl = map["photoUrl"] as? String ?: "",
                senderId = map["senderId"] as? String ?: "",
                senderName = map["senderName"] as? String ?: "",
                senderPhoto = map["senderPhoto"] as? String ?: "",
                createdAtMillis = (map["createdAt"] as? Timestamp)?.toDate()?.time ?: 0L,
                replyTo = reply
            )
        }
    }
}
